### PR TITLE
Add support for a symbolic output_shape for broadcast_in_dim

### PR DESCRIPTION
**Pull request by kevinstephano**
Augmenting Ivan's code: IvanYashchuk@7f2437a.

Adding a new arith function to get the `Val`s for each `TensorView` dimension `extent` and passing them to `broadcast_in_dim`.

I needed to modify the `nvfuser::Tensor` objects to include an attribute `dims` in order to propagate around the number of dimensions of each tensor.

Example:

```
def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1, -1, -1], contiguous=[False, False, False], dtype=DataType.Float, is_cpu=False)
    T1 = fd.define_tensor(symbolic_sizes=[-1], contiguous=[False], dtype=DataType.Float, is_cpu=False)
    S2, S3, S4 = fd.ops.tensor_sizes(T0)
    T5 = fd.ops.broadcast_in_dim(T1, output_shape=[S2, S3, S4], broadcast_dims=[2])
    T6 = fd.ops.add(T0, T5)
    fd.add_output(T6)
```

Todo:
- [x] Fix the compilation of the `BroadcastInDimOpRecord::operator()` that needs to use template specialization to handle type specific logic.
- [x] Add a SymbolicSizesRecord to represent the new arith function.
- [x] Add python binding for `tensor_sizes()` operator.
- [x] Add tests
- [x] Add comments

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/arith.h [hunk 1]
```diff
@@ -364,6 +364,8 @@ TORCH_CUDA_CU_API TensorView* expand(
 // non broadcasted iter domain, inp will be expanded to other's size.
 TORCH_CUDA_CU_API TensorView* expand_as(TensorView* inp, TensorView* other);
 
+TORCH_CUDA_CU_API std::vector<Val*> symbolic_sizes(TensorView* inp);
```
Comment (IvanYashchuk) [inline torch/csrc/jit/codegen/cuda/arith.h]: Should we take as a guidance ATen C++ naming or PyTorch naming?
In PyTorch, the shape/size of a tensor is obtained with `tensor.shape` or `tensor.size()`.
In ATen C++ the shape/size of a tensor is obtained with `tensor.sizes()`.

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 2]
```diff
@@ -1616,6 +1616,17 @@ TensorView* expand_as(TensorView* inp, TensorView* other) {
   return out_tensor;
 }
 
+std::vector<Val*> symbolic_sizes(TensorView* inp) {
+  auto iter_domains = inp->domain()->noReductions(inp->getMaybeRFactorDomain());
```
Comment (IvanYashchuk) [inline torch/csrc/jit/codegen/cuda/arith.cpp]: Another alternative to do the same, seems clearer to me:
```suggestion
  auto iter_domains = TensorDomain::noReductions(inp->getMaybeRFactorDomain());
```

File: torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h [hunk 3]
```diff
@@ -692,13 +667,56 @@ struct BroadcastInDimOpRecord : RecordFunctor {
...
   //! For instance, for output [2, 3, 4] and input [3]. This vector would
   //! contain [1].
   std::vector<int64_t> broadcast_dims_;
 };
 
+template<>
+inline c10::optional<std::vector<Nvf::Val*>>
+BroadcastInDimOpRecord<int64_t>::expandShape(
+    const FusionDefinition& fd,
+    const std::vector<bool> expand_dim,
```
Comment (IvanYashchuk) [inline torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h]:
```suggestion
    const std::vector<bool>& expand_dim,
```

File: torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h [hunk 4]
```diff
@@ -602,6 +585,11 @@ struct BroadcastInDimOpRecord : RecordFunctor {
     return result;
   }
 
+  inline c10::optional<std::vector<Nvf::Val*>>
+  expandShape(const FusionDefinition& fd,
+              const std::vector<bool> expand_dim,
```
Comment (IvanYashchuk) [inline torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h]:
```suggestion
              const std::vector<bool>& expand_dim,
```

File: torch/csrc/jit/codegen/cuda/arith.h (hunk 1 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/arith.h]: I changed it to `tensor_sizes()`.

File: torch/csrc/jit/codegen/cuda/python_frontend/fusion_definition.cpp [hunk 5]
```diff
@@ -118,7 +141,7 @@ void FusionDefinition::print(std::ostream& os) const {
     rec->print(os);
     os << "\n";
   }
-  os << "\n";
+  os << "\n" << std::flush;
```
Comment (mruberry) [inline torch/csrc/jit/codegen/cuda/python_frontend/fusion_definition.cpp]: `std::endl`? https://en.cppreference.com/w/cpp/io/manip/endl

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 6]
```diff
@@ -1667,6 +1667,17 @@ TensorView* expand_as(TensorView* inp, TensorView* other) {
   return out_tensor;
 }
 
+std::vector<Val*> tensor_sizes(TensorView* inp) {
+  auto iter_domains = TensorDomain::noReductions(inp->getMaybeRFactorDomain());
+  std::vector<Val*> sizes(iter_domains.size(), nullptr);
+
+  for (auto idx : c10::irange(iter_domains.size())) {
+    sizes[idx] = iter_domains[idx]->extent();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/arith.cpp]: I think this should actually be `getMaybeExpandedExtent`. You could probably test this by doing an expand then using the expanded size for a broadcast. If we don't use `getMaybeExpandedExtent` the extent returned would be compile time constant value 1.

**Review by csarofeen — APPROVED** [review]
LGTM though I think it needs one minor fix. Stamping as approved but please check my comment first.

File: torch/csrc/jit/codegen/cuda/python_frontend/fusion_definition.cpp (hunk 5 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/python_frontend/fusion_definition.cpp]: Changed, thanks!

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 6 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: You were right.  I added the appropriate fix and a test.

**Pull request merged**